flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java

```diff
@@ -195,12 +216,18 @@ public void processWatermark(Watermark mark) throws Exception {
 
     @Override
     public void endInput() throws Exception {
-        endOfInput = true;
-        sinkWriter.flush(true);
-        emitCommittables(Long.MAX_VALUE);
+        if (!this.endOfInput) {
+            this.endOfInput = true;
+            // endOfInputState is union state, so it's enough if one task adds something to it
+            if (getRuntimeContext().getTaskInfo().getIndexOfThisSubtask() == 0) {
+                endOfInputState.add(true);
```

Comment:
What happens if we up-scale after this? 
should we drop all other states?
Otherwise, should we check in initState() all items?

Reply:
I changed the logic entirely. We now drop and readd the state in `snapshotState`. This means if a downscaling happens and EOI_state = [true, false], on init endOfInput = false, and on next checkpoint EOI_state=[false]. I discussed the different cases in code.